File: packages/py/gof_patterns/structural.py
"""Structural patterns.

Three need no helper — see the README table:

- Flyweight -> ``functools.lru_cache`` (shared instances per key, plus
  ``cache_clear`` and ``cache_info`` for free)
- Facade    -> an object that delegates. Just write it.
- Bridge    -> pass the implementation in. Just write it.
"""

from __future__ import annotations

import functools
from types import SimpleNamespace
from typing import Any, Callable, Generic, Iterator, TypeVar

T = TypeVar("T")


def adapt(source: Any, **methods: Callable[[Any], Any]) -> SimpleNamespace:
    """Adapter — expose an incompatible object through the interface you want.

        api = adapt(legacy, log=lambda l: l.send_message)
        api.log("hi")
    """
    return SimpleNamespace(**{name: build(source) for name, build in methods.items()})


def decorate(fn: Callable[..., Any], *wrappers: Callable[..., Any]) -> Callable[..., Any]:
    """Decorator — wrap a callable in layers. The first wrapper is outermost.

        fetch_user = decorate(raw_fetch_user, with_log, with_retry(3))

    Same composition ``@with_log`` gives you, but applied to an existing object
    at runtime and readable left-to-right.
    """
    wrapped = fn
    for wrapper in reversed(wrappers):
        wrapped = functools.wraps(fn)(wrapper(wrapped))
    return wrapped


class Composite(Generic[T]):
    """Composite — treat a tree of nodes like a single node.

        box = Composite({"price": 10})
        box.add(Composite({"price": 5}))
        box.sum(lambda v: v["price"])  # 15
    """

    def __init__(self, value: T, children: list["Composite[T]"] | None = None) -> None:
        self.value = value
        self.children: list[Composite[T]] = children if children is not None else []

    def add(self, *children: "Composite[T]") -> "Composite[T]":
        self.children.extend(children)
        return self

    def remove(self, child: "Composite[T]") -> bool:
        try:
            self.children.remove(child)
        except ValueError:
            return False
        return True

    def walk(self) -> Iterator["Composite[T]"]:
        """Depth-first, self first."""
        yield self
        for child in self.children:
            yield from child.walk()

    def sum(self, of: Callable[[T], float]) -> float:
        return sum(of(node.value) for node in self.walk())

    def __len__(self) -> int:
        return sum(1 for _ in self.walk())


class lazy(Generic[T]):  # noqa: N801 - reads as a function at the call site
    """Proxy (virtual) — build the real object on first attribute access.

        heavy = lazy(load_huge_thing)  # nothing loaded yet
        heavy.query("x")               # loads now, once
    """

    __slots__ = ("_loader", "_target")

    def __init__(self, loader: Callable[[], T]) -> None:
        object.__setattr__(self, "_loader", loader)
        object.__setattr__(self, "_target", None)

    def _resolve(self) -> T:
        target = object.__getattribute__(self, "_target")
        if target is None:
            target = object.__getattribute__(self, "_loader")()
            object.__setattr__(self, "_target", target)
        return target

    def __getattr__(self, name: str) -> Any:
        return getattr(self._resolve(), name)

    def __setattr__(self, name: str, value: Any) -> None:
        setattr(self._resolve(), name, value)

File: packages/ts/src/structural.ts
/** Structural patterns. */

/**
 * Adapter — expose an incompatible object through the interface you want.
 *
 * ```ts
 * const logger = adapt(winston, { log: (w) => (m: string) => w.info(m) });
 * ```
 */
export function adapt<S, A extends Record<string, (source: S) => unknown>>(
  source: S,
  methods: A,
): { [K in keyof A]: ReturnType<A[K]> } {
  const out = {} as { [K in keyof A]: ReturnType<A[K]> };
  for (const key of Object.keys(methods) as (keyof A)[]) {
    out[key] = methods[key](source) as ReturnType<A[keyof A]>;
  }
  return out;
}

/**
 * Decorator — wrap a function in layers (retry, cache, log) without touching it.
 * Applied left to right: the first wrapper is the outermost.
 *
 * ```ts
 * const fetchUser = decorate(rawFetchUser, withLog, withRetry(3));
 * ```
 */
export type Wrapper<F extends (...args: never[]) => unknown> = (next: F) => F;

export function decorate<F extends (...args: never[]) => unknown>(
  fn: F,
  ...wrappers: Wrapper<F>[]
): F {
  return wrappers.reduceRight((next, wrap) => wrap(next), fn);
}

/**
 * Composite — treat a tree of nodes like a single node.
 *
 * ```ts
 * const box = composite({ price: 10 });
 * box.add(composite({ price: 5 }));
 * box.sum((n) => n.price); // 15
 * ```
 */
export type Composite<T> = {
  readonly value: T;
  readonly children: Composite<T>[];
  add(...children: Composite<T>[]): Composite<T>;
  remove(child: Composite<T>): boolean;
  /** Depth-first, self first. */
  walk(): Generator<Composite<T>>;
  sum(of: (value: T) => number): number;
};

export function composite<T>(value: T, children: Composite<T>[] = []): Composite<T> {
  const node: Composite<T> = {
    value,
    children,
    add(...kids) {
      children.push(...kids);
      return node;
    },
    remove(child) {
      const i = children.indexOf(child);
      if (i < 0) return false;
      children.splice(i, 1);
      return true;
    },
    *walk() {
      yield node;
      for (const child of children) yield* child.walk();
    },
    sum(of) {
      let total = 0;
      for (const n of node.walk()) total += of(n.value);
      return total;
    },
  };
  return node;
}

/**
 * Flyweight — share immutable instances instead of re-creating them.
 *
 * ```ts
 * const treeType = flyweight((name: string, color: string) => ({ name, color }));
 * treeType('oak', 'green') === treeType('oak', 'green'); // true
 * ```
 * Default key is `JSON.stringify(args)`; pass `key` for anything richer.
 */
export function flyweight<A extends unknown[], T>(
  factory: (...args: A) => T,
  key: (...args: A) => string = (...args) => JSON.stringify(args),
): ((...args: A) => T) & { size(): number; clear(): void } {
  const cache = new Map<string, T>();
  const get = (...args: A): T => {
    const k = key(...args);
    if (!cache.has(k)) cache.set(k, factory(...args));
    return cache.get(k) as T;
  };
  get.size = () => cache.size;
  get.clear = () => cache.clear();
  return get;
}

/**
 * Proxy (virtual) — build the real object on first property access.
 *
 * ```ts
 * const heavy = lazy(() => loadHugeThing()); // nothing loaded yet
 * heavy.query('x');                          // loads now, once
 * ```
 */
export function lazy<T extends object>(loader: () => T): T {
  let target: T | undefined;
  const get = () => (target ??= loader());
  return new Proxy({} as T, {
    get: (_t, prop, receiver) => Reflect.get(get(), prop, receiver),
    set: (_t, prop, value) => Reflect.set(get(), prop, value),
    has: (_t, prop) => Reflect.has(get(), prop),
    ownKeys: () => Reflect.ownKeys(get()),
    getOwnPropertyDescriptor: (_t, prop) => Reflect.getOwnPropertyDescriptor(get(), prop),
  });
}

// ponytail: Facade and Bridge are plain composition — an object or class that
// delegates. No runtime helper can make that shorter than writing it. See README.
